chromadb/utils/embedding_functions.py

```diff
@@ -340,46 +340,45 @@ def _forward(self, documents: List[str], batch_size: int = 32) -> npt.NDArray:
             all_embeddings.append(embeddings)
         return np.concatenate(all_embeddings)
 
-    def _init_model_and_tokenizer(self) -> None:
-        if self.model is None and self.tokenizer is None:
-            self.tokenizer = self.Tokenizer.from_file(
-                os.path.join(
-                    self.DOWNLOAD_PATH, self.EXTRACTED_FOLDER_NAME, "tokenizer.json"
-                )
+    @cached_property
+    def tokenizer(self) -> "Tokenizer":
```

Comment:
Aren't race conditions still possible even if this is cached?

Reply:
Ostensibly they are still possible but they are harmless, at worst the model or the tokenizer are initialized more than once but this is idempotent.
The reason the current race condition is problematic is the `and` in `if self.model is None and self.tokenizer is None`:
1. Initially the condition is True as both model and tokenizer are None
2. Thread A enters the block and initializes the tokenizer
3. The context switches to thread B before A gets to initialize the model
4. Thread B evaluates the condition as False (model is None, tokenizer is not None)
5. Thread B attempts to access `self.model.run` 
6. `AttributeError` 💥

Reply:
Fair enough :) Thanks for the analysis!